### PR TITLE
 Support for regex-based ignore

### PR DESCRIPTION
**Pull request by arusahni**
This addresses issue #156 by allowing for multiple ignore patterns
to be specified. Rather than clobber the existing ignore option, I've
introduced a new one called ignore-patterns. That way there are as few
surprises as something involving regular expressions can have.

Multiple patterns must be separated by a comma. For example, if I
wanted to ignore all files that started with `test_` and `_`, I'd pass this
in: `--ignore-patterns="test_.*","_.*"`.

### CODE REVIEW

File: pylint/config.py [hunk 1]
```diff
@@ -162,6 +162,9 @@ def _regexp_validator(_, name, value):
         return value
     return re.compile(value)
 
+# pylint: disable=unused-argument
```
Comment (PCManticore) [inline pylint/config.py]: What argument is not used? It seems that both name and value are used and the first one is marked as unused anyway with underscore.

**Comment by PCManticore** [comment]
Thank you for the PR! And sorry for taking so long to review it, it was a hectic month for me.

Would you like to try to add a couple of tests? Except this, everything looks pretty solid.

**Comment by arusahni** [comment]
I took out the unused argument disable - it was a holdover from an earlier version that I had written.  I also added some tests for the validators and file matching components.

**Comment by PCManticore** [comment]
I merged this locally and squashed your commits into one: https://github.com/PyCQA/pylint/commit/2ec342e612c730a266413783918f4a0d41441ca5

**Comment by arusahni** [comment]
@PCManticore thanks!

**Comment by appierys** [comment]
@arusahni Just curious. Why don't we support regex for paths but only for basenames?
In #729, he originally wanted the support for `paths` rather than `basenames`, right?
Are any reasons that this is discouraged?

**Comment by arusahni** [comment]
@appierys I chose to have this only operate with basenames as that was the closest analog to the existing `--ignore` behavior. In retrospect this may not have addressed the original ticket, however.  I may have misread and confused the "regex" portion of the ticket thread as being satisfied by my work.

Feel free to comment on that ticket to see if it's worth reopening.

**Pull request closed without merging**